clusterloader2/pkg/util/cluster.go

```diff
@@ -0,0 +1,179 @@
+	taintNodeNotReady    = "node.kubernetes.io/not-ready"
+	taintNodeUnreachable = "node.kubernetes.io/unreachable"
+)
+
+var (
+	// UnreachableTaintTemplate is the taint for when a node becomes unreachable.
+	unreachableTaintTemplate = &corev1.Taint{
+		Key:    taintNodeUnreachable,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+
+	// NotReadyTaintTemplate is the taint for when a node is not ready for
+	// executing pods
+	notReadyTaintTemplate = &corev1.Taint{
+		Key:    taintNodeNotReady,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+)
+
+// GetNodesNumber returns number of nodes in the cluster.
+func GetNodesNumber(c clientset.Interface) (int, error) {
+	nodeList, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return 0, err
+	}
+	numNodes := 0
+	for i := range nodeList.Items {
+		if isNodeSchedulable(&nodeList.Items[i]) && isNodeUntainted(&nodeList.Items[i]) {
+			numNodes++
+		}
+	}
+	return numNodes, err
+}
+
+// Node is schedulable if:
+// 1) doesn't have "unschedulable" field set
+// 2) it's Ready condition is set to true
+// 3) doesn't have NetworkUnavailable condition set to true
+func isNodeSchedulable(node *corev1.Node) bool {
+	nodeReady := isNodeConditionSetAsExpected(node, corev1.NodeReady, true, false)
+	networkReady := isNodeConditionUnset(node, corev1.NodeNetworkUnavailable) ||
+		isNodeConditionSetAsExpected(node, corev1.NodeNetworkUnavailable, false, true)
+	return !node.Spec.Unschedulable && nodeReady && networkReady
+}
+
+// Test whether a fake pod can be scheduled on "node", given its current taints.
+func isNodeUntainted(node *corev1.Node) bool {
+	fakePod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-not-scheduled",
+			Namespace: "fake-not-scheduled",
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:  "fake-not-scheduled",
+					Image: "fake-not-scheduled",
+				},
+			},
+		},
+	}
+
+	fit := podToleratesNodeTaints(fakePod, node, func(t *corev1.Taint) bool {
+		// PodToleratesNodeTaints is only interested in NoSchedule and NoExecute taints.
+		return t.Effect == corev1.TaintEffectNoSchedule || t.Effect == corev1.TaintEffectNoExecute
+	})
+	return fit
+}
+
+func podToleratesNodeTaints(pod *corev1.Pod, node *corev1.Node, filter func(t *corev1.Taint) bool) bool {
+	if v1helper.TolerationsTolerateTaintsWithFilter(pod.Spec.Tolerations, node.Spec.Taints, filter) {
+		return true
+	}
+	return false
+}
+
+func isNodeConditionSetAsExpected(node *corev1.Node, conditionType corev1.NodeConditionType, wantTrue, silent bool) bool {
+	// Check the node readiness condition (logging all).
+	for _, cond := range node.Status.Conditions {
+		// Ensure that the condition type and the status matches as desired.
+		if cond.Type == conditionType {
+			// For NodeReady condition we need to check Taints as well
+			if cond.Type == corev1.NodeReady {
```

Comment:
I would really like to simplify this function.
In particular:

1. do we really need to check taints? we are checking those separately in isNodeUntainted anyway.
Or do I miss something?

2. Probably more, I think. But let's ensure that taints aren't really needed and that would simplify stuff..

Reply:
If I understand isNodeUntainted correctly, it will return false if any of taints has TaintEffectNoExecute effect. Assuming that unreachableTaintTemplate and notReadyTaintTemplate also have the TaintEffectNoExecute effect, this case will be verified by isNodeUntainted.

I agree that taints check can be removed.

Reply:
After removing taint part, it turned out that the whole `NodeReady` case check is not needed.